### ISSUE TITLE
[Flags] [VariableUtils] SetFlag not working with NOT_FLAG

### ISSUE THREAD

**Issue by miguelmaso** [KratosMultiphysics/Kratos#4896]
From the python interface, `node.Set(NOT_Flag)` is working, but `VariableUtils.SetFlag(NOT_Flag, True, nodes_container)` does nothing.

This test file will produce an error:
~~~py
import KratosMultiphysics as KM
import KratosMultiphysics.KratosUnittest as KratosUnittest

class TestVariableUtilsSetNotFlag(KratosUnittest.TestCase):
    def test_set_not_flag(self):
        model = KM.Model()
        model_part = model.CreateModelPart("model_part")
        node = model_part.CreateNewNode(1, 0.0, 0.0, 0.0)

        self.assertFalse(node.Is(KM.INTERFACE))
        self.assertTrue(node.Is(KM.NOT_INTERFACE))

        KM.VariableUtils().SetFlag(KM.INTERFACE, True, model_part.Nodes)
        self.assertTrue(node.Is(KM.INTERFACE))

        KM.VariableUtils().SetFlag(KM.NOT_INTERFACE, True, model_part.Nodes)
        self.assertTrue(node.Is(KM.NOT_INTERFACE))

if __name__ == '__main__':
    KratosUnittest.main()

~~~

I think this is not the expected behavior

**Comment by miguelmaso** [comment]
Using:
~~~py
KM.VariableUtils().SetFlag(KM.INTERFACE, False, model_part.Nodes)
~~~
the test will pass

**Comment by loumalouomega** [comment]
Hummm, in theory the NOT is just a flag, should do the same...

**Comment by jcotela** [comment]
Actually I am afraid it is expected behavior :man_facepalming: 
`FLAG` and `NOT_FLAG` both refer to the same flag, in one case set to `true` and in the other set to `false`, so that, for example:
```python
if node.Is(STRUCTURE):
   print("The node has the STRUCTURE flag set to True")

if node.Is(NOT_STRUCTURE):
   print("The node has the STRUCTURE flag set to False")
```
So, when you do
```python
node.Set(NOT_STRUCTURE, True)
```
it is understood by Kratos as "get the bit storing the `STRUCTURE` flag and set it to `true`", which is exactly what it is doing.

I believe that we may be able to use some trickery to make it so
```python
node.Set(NOT_STRUCTURE,True)
```
is equivalent to
```python
node.Set(STRUCTURE,False)
```
but then I do not want to event start thinking of what should be the result (and the implementation) of
```python
node.Set(STRUCTURE | NOT_INLET,False)
```

**Comment by miguelmaso** [comment]
> I believe that we may be able to use some trickery to make it so
> 
> ```python
> node.Set(NOT_STRUCTURE,True)
> ```
> 
> is equivalent to
> 
> ```python
> node.Set(STRUCTURE,False)
> ```
I have tested it and it seems to work fine. The problem I found is in VariableUtils, this test is more clear:
~~~py
import KratosMultiphysics as KM
import KratosMultiphysics.KratosUnittest as KratosUnittest

class TestVariableUtilsSetNotFlag(KratosUnittest.TestCase):
    def test_set_not_flag(self):
        model = KM.Model()
        model_part = model.CreateModelPart("model_part")
        node = model_part.CreateNewNode(1, 0.0, 0.0, 0.0)

        # 1. It works fine
        node.Set(KM.INTERFACE)
        node.Set(KM.NOT_INTERFACE)
        self.assertTrue(node.Is(KM.NOT_INTERFACE))

        # 2. This should be the same as 1, but VariableUtils has an unexpected behavior
        node.Set(KM.INTERFACE)
        KM.VariableUtils().SetFlag(KM.NOT_INTERFACE, True, model_part.Nodes)
        self.assertTrue(node.Is(KM.NOT_INTERFACE))

if __name__ == '__main__':
    KratosUnittest.main()
~~~

The point is the inconsistency between Set() and VariableUtils().SetFlag()

**Comment by jcotela** [comment]
What I am trying to say is that `VariableUtils().SetFlag(FLAG,Value)` only calls `node.Set(FLAG,Value)` internally. It is consistent, even if its internal logic is *very* counterintuitive. 

Please allow me a detour into the internals of the `Flags` (and excuse me if you know this already, but I've found that even experienced developers of Kratos are not aware of it -- I also wasn't before I had to dig into it for MPI synchronization):

`Kratos::Flags` is a collection of bits. Each individual named flag is identified by two of these bits, one of the bits tells us if the flag is defined (1 if yes, 0 if no), the other tells us if its value is true (1) or false (0). Named Flags are only aliases for these bit pairs.

For example `INTERFACE` is a named `Kratos::Flags` object with one of these pairs of bits set to a predefined value (this is the pair of bits that we "label" with the name `INTERFACE`). Specifically, one of the "defined" bit is 1 (flag is defined) and its corresponding "value" bit is 1 (its value is true).

Similarly, `NOT_INTERFACE` is another named `Kratos::Flags` object where the same pair of bits has been set. Here, the **same** "defined" bit is 1 but the corresponding "value" bit is 0.

With this in mind, lets go back to the example:
```python
import KratosMultiphysics as KM
import KratosMultiphysics.KratosUnittest as KratosUnittest

class TestVariableUtilsSetNotFlag(KratosUnittest.TestCase):
    def test_set_flags(self):
        model = KM.Model()
        model_part = model.CreateModelPart("model_part")
        node = model_part.CreateNewNode(1, 0.0, 0.0, 0.0)

        # 1. set the INTERFACE bits of the node as in the named flag KM.INTERFACE
        node.Set(KM.INTERFACE)
        self.assertTrue(node.IsDefined(KM.INTERFACE)) # defined
        self.assertTrue(node.Is(KM.INTERFACE)) # and its value is true

        # 2. set the INTERFACE bits of the node as in the named flag KM.NOT_INTERFACE
        node.Set(KM.NOT_INTERFACE)
        self.assertTrue(node.IsDefined(KM.INTERFACE)) # defined
        self.assertTrue(node.IsDefined(KM.NOT_INTERFACE)) # defined (it is really the same check as in the line above)
        self.assertFalse(node.Is(KM.INTERFACE)) # and its value is false

        # 3. set the value bit for INTERFACE to True
        # (when setting a flag in this way, the corresponding defined bit is always set to 1 internally)
        node.Set(KM.INTERFACE, True)
        self.assertTrue(node.IsDefined(KM.INTERFACE)) # defined
        self.assertTrue(node.Is(KM.INTERFACE)) # and its value is true

        # 4. Also set the INTERFACE value bit of the node to True 
        # (here I use the named object KM.NOT_INTERFACE as reference, which points to the same pair of bits as INTERFACE)
        node.Set(KM.NOT_INTERFACE, True)
        self.assertTrue(node.IsDefined(KM.INTERFACE)) # defined
        self.assertTrue(node.IsDefined(KM.NOT_INTERFACE)) # defined (again, the same check as in the line above)
        self.assertTrue(node.Is(KM.INTERFACE)) # and its value is true

        # 5. VariableUtils calls the same method as 4, internally
        node.Set(KM.INTERFACE)
        KM.VariableUtils().SetFlag(KM.NOT_INTERFACE, True, model_part.Nodes)
        self.assertTrue(node.Is(KM.INTERFACE))

if __name__ == '__main__':
    KratosUnittest.main()
```

So my recommendation would be to make VariableUtils use the same method as 1 above (instead of the method of 3). This means that it does not accept a `bool` argument. The flag is set to a true/false value depending on whether we got `FLAG` or `NOT_FLAG` as argument. This also makes it possible to do things like `KM.VariableUtils().SetFlag(KM.INLET | KM.NOT_OUTLET, model_part.Nodes)` to set two flags at the same time.

If you ask me, the method used to set the flags in options 3, 4 and (internally) 5 should be removed, because it allows you to do things that do not make intuitive sense, as this issue shows.

**Comment by miguelmaso** [comment]
Thanks for the explanation

**Comment by philbucher** [comment]
can this be closed?

**Comment by jcotela** [comment]
@philbucher the facts are still the same: `KM.VariableUtils().SetFlag(KM.NOT_INTERFACE, True, model_part.Nodes)` sets `INTERFACE`, so `NOT_INTERFACE` is `False` after the call. I think we all can agree that this makes no sense.

Should we:
1. Close this and forget about it?
2. Change the behavior of the `Flags` to make them more logical? (If so, how?)
3. Remove `KM.VariableUtils().SetFlag(KM.NOT_INTERFACE, True, model_part.Nodes)` and remove the `bool` argument, so that true/false is determined by the value of the flag argument? (This would set the flag to `true` when setting `INTERFACE` and to `false` when setting `NOT_INTERFACE`).

In any of these cases, what happens when setting multiple flags on the same call? (for example, what does `KM.VariableUtils().SetFlag(KM.NOT_INTERFACE | KM.OUTLET, True, model_part.Nodes)` mean, precisely?).

Personally, I would go for 3. above (which would be an API breaker, unfortunately), but I would encourage the @KratosMultiphysics/technical-committee to think what the desired behavior should be.

**Comment by philbucher** [comment]
thanks @jcotela for the explanation, I added it to the committee to be discussed

**Comment by GuillermoCasas** [comment]
I definitely agree with the proposed simplification by @jcotela. The behaviour shown in the example # 5 above is very confusing and I think only the result of the limitations of the Flags object design. Therefore, we should remove the possibility of 'setting' any flag to the value of a bool; at least until the behaviour is fixed (which is probably not easy to do).

**Comment by AlejandroCornejo** [comment]
The @KratosMultiphysics/implementation-committee agrees with @jcotela and @GuillermoCasas. Furthermore, what is really the need of having the "NOT" version of every flag? Could we not just remove them? 
Probably tha latter question deserves a separate issue.

**Comment by pooyan-dadvand** [comment]
Just for the record, I had a discussion about this with @RiccardoRossi and in our opinion, we might 
1. Remove the NOT version of the flags. (There are no many uses of them)
2. Removing the is defined also to allocate less memory and use a separate flag for the definition of them.

**Comment by loumalouomega** [comment]
> 
> 
> Just for the record, I had a discussion about this with @RiccardoRossi and in our opinion, we might
> 
>     1. Remove the NOT version of the flags. (There are no many uses of them)
> 
>     2. Removing the is defined also to allocate less memory and use a separate flag for the definition of them.

This means that if we remove the NOT we will be able to register the double number of flags?

**Comment by jcotela** [comment]
@loumalouomega having the NOT does not require extra memory, but having the DEFINED bit does, so the memory use should be halved.

**Comment by roigcarlo** [comment]
@KratosMultiphysics/technical-committee agrees on taking care of this in two steps. The first one will be to remove the NOT from the flags as it is rarely used and misleading (@rubenzorrilla will take care of this one). In the second one remove the DEFINE, which implies a memory consumption reduction (@jcotela could you please take care of the second step?).

**Comment by jcotela** [comment]
@KratosMultiphysics/technical-committee what happens with current checks for IsDefined? I would not want to take care of all of them in a single PR. In fact, I would shoot on sight any PR that mixes changing the current behavior of the `ACTIVE` flag with anything else.

**Comment by pooyan-dadvand** [comment]
You are right! We should first deprecate the `IsDefined` and ask all the applications to clean the warning before you take it out.

**Comment by loumalouomega** [comment]
> 
> 
> You are right! We should first deprecate the `IsDefined` and ask all the applications to clean the warning before you take it out.

This is going to be harmful

**Comment by pooyan-dadvand** [comment]
Yes, there are many (about 500) but not too much

**Comment by loumalouomega** [comment]
> 
> 
> Yes, there are many (about 500) but not too much

What I am scared is that I remember that at some point i considered the IsDefined as a "third flag option", and some of my implementations use that in order to discard options, so we must be specially careful (at least in my side I have 100+ tests)

**Comment by pooyan-dadvand** [comment]
😨

**Comment by rubenzorrilla** [comment]
Hey! While removing the `NOT_` from the flags I found a problem when we set defaults in constructors/methods. Below you can see an example of this situation. I tried by doing `!FLAG` but it doesn't compile (something related to the conversion from `Flag` to `int`). Any hint? @pooyan-dadvand 

```c++
    FindIntersectedGeometricalObjectsWithOBBContactSearchProcess(
        ModelPart& rPart1,
        ModelPart& rPart2,
        const double BoundingBoxFactor = -1.0,
        const Flags Options = BaseProcessType::INTERSECTING_CONDITIONS|BaseProcessType::INTERSECTING_ELEMENTS|BaseProcessType::INTERSECTED_CONDITIONS|BaseProcessType::INTERSECTED_ELEMENTS|BaseType::NOT_DEBUG_OBB|BaseType::SEPARATING_AXIS_THEOREM|BaseType::BUILD_OBB_FROM_BB
        );
```

**Comment by jcotela** [comment]
@rubenzorrilla I think you should be able to use `~FLAG`

**Comment by rubenzorrilla** [comment]
> @rubenzorrilla I think you should be able to use `~FLAG`

Cool. I'll try. Actually we were thinking about defining an `!` operator for the flags.

**Comment by rubenzorrilla** [comment]
After #6233 there are no `NOT_FLAGS` any more. Closing the issue. @miguelmaso please feel free to reopen it if required.

**Comment by miguelmaso** [comment]
Thanks @rubenzorrilla
